ledger/src/shred/merkle.rs

```diff
@@ -154,9 +254,66 @@ impl ShredCode {
             || self.merkle_branch.root != other.merkle_branch.root
             || self.common_header.signature != other.common_header.signature
     }
+
+    fn from_recovered_shard(
+        common_header: ShredCommonHeader,
+        coding_header: CodingShredHeader,
+        mut shard: Vec<u8>,
+    ) -> Result<Self, Error> {
+        let proof_size = match common_header.shred_variant {
+            ShredVariant::MerkleCode(proof_size) => proof_size,
+            _ => return Err(Error::InvalidShredVariant),
+        };
+        let shard_size = shard.len();
+        if Self::capacity(proof_size)? != shard_size {
+            return Err(Error::InvalidShardSize(shard_size));
+        }
+        if shard_size + Self::SIZE_OF_HEADERS > Self::SIZE_OF_PAYLOAD {
+            return Err(Error::InvalidShardSize(shard_size));
+        }
+        shard.resize(Self::SIZE_OF_PAYLOAD, 0u8);
+        shard.copy_within(0..shard_size, Self::SIZE_OF_HEADERS);
+        let mut cursor = Cursor::new(&mut shard[..]);
+        bincode::serialize_into(&mut cursor, &common_header)?;
+        bincode::serialize_into(&mut cursor, &coding_header)?;
+        Ok(Self {
+            common_header,
+            coding_header,
+            merkle_branch: MerkleBranch::new_zeroed(proof_size),
+            payload: shard,
+        })
+    }
+
+    fn set_merkle_branch(&mut self, merkle_branch: MerkleBranch) -> Result<(), Error> {
+        let proof_size = self.proof_size()?;
+        if merkle_branch.proof.len() != usize::from(proof_size) {
+            return Err(Error::InvalidMerkleProof);
+        }
+        let offset = Self::SIZE_OF_HEADERS + Self::capacity(proof_size)?;
+        let mut cursor = Cursor::new(
+            self.payload
+                .get_mut(offset..)
+                .ok_or(Error::InvalidProofSize(proof_size))?,
+        );
+        bincode::serialize_into(&mut cursor, &merkle_branch.root)?;
```

Comment:
create a utility function to encode `MerkleBranch`?

Reply:
Keeping this inline for now to be explicit.
This serialization relies on `merkle_branch.proof.len()` being encoded out-of-band so it cannot serve as a standalone encoding/serialization method for `MerkleBranch`.

Reply:
ok